### PR TITLE
create a miri-pass test that allows us to run miri for arbitrary targets

### PR DESCRIPTION
**Pull request by oli-obk**
compiletest doesn't give us this kind of pass, so I created one...

now I can miri-run windows-tests on linux:

```
running tests for target x86_64-pc-windows-gnu
test [miri-pass] tests/run-pass/loops.rs
test [miri-pass] tests/run-pass/arrays.rs
```

The test suite automatically checks for installed targets and runs miri on them.

Also I hope the travis changes now actually break the CI if the tests fail.

I needed compiletest 2.0 for `compiletest::procsrv::dylib_env_var()`

### CODE REVIEW

File: tests/compiletest.rs [hunk 1]
```diff
@@ -1,43 +1,73 @@
...
+            writeln!(stderr.lock(), "test [miri-pass] {}", file.to_str().unwrap()).unwrap();
+            let mut cmd = std::process::Command::new("target/debug/miri");
+            cmd.arg(file);
+            cmd.arg(format!("--sysroot={}", sysroot));
+            cmd.arg("-Dwarnings");
+            cmd.arg(format!("-target={}", target));
+            let libs = Path::new(&sysroot).join("lib");
+            let sysroot = libs.join("rustlib").join(&target).join("lib");
+            let paths = std::env::join_paths(&[libs, sysroot]).unwrap();
+            cmd.env(compiletest::procsrv::dylib_env_var(), paths);
```
Comment (solson) [inline tests/compiletest.rs]: It seems like this `cmd` never gets executed, only args and env are set.

**Comment by solson** [comment]
Travis failed:

> thread 'compile_test' panicked at 'need to specify RUST_SYSROOT env var or use rustup or multirust', ../src/libcore/option.rs:699

File: .travis.yml [hunk 2]
```diff
@@ -9,9 +9,10 @@ before_script:
   pip install 'travis-cargo<0.2' --user &&
   export PATH=$HOME/.local/bin:$PATH
 script:
-- |
-  travis-cargo build &&
-  env RUST_SYSROOT=$HOME/rust travis-cargo test
+- set -e
+- travis-cargo build
+- RUST_SYSROOT=$HOME/rust
```
Comment (solson) [inline .travis.yml]: I think this needs an `export` if you do it as a separate command. This probably explains the Travis failure on this PR.

File: tests/compiletest.rs [hunk 3]
```diff
@@ -1,43 +1,73 @@
...
+    });
+}
+
+fn for_all_targets<F: Fn(String)>(sysroot: &str, f: F) {
+    for target in std::fs::read_dir(format!("{}/lib/rustlib/", sysroot)).unwrap() {
+        let target = target.unwrap();
+        if !target.metadata().unwrap().is_dir() {
+            continue;
+        }
+        let target = target.path().iter().rev().next().unwrap().to_str().unwrap().to_owned();
```
Comment (solson) [inline tests/compiletest.rs]: You can use `DirEntry::file_name` instead:

```rust
let target = target.file_name().into_string().unwrap();
```

Comment (oli-obk) [inline tests/compiletest.rs]: makes sense

**Comment by oli-obk** [comment]
out of some reason travis still doesn't fail when `travis-cargo test` fails...

**Comment by oli-obk** [comment]
now it works as it should.

**Comment by solson** [comment]
Thanks a ton for figuring all that stuff out!

**Pull request merged**